File: runtime/jvm/test/lox/FileTest.java
package lox;

import static lox.TestSupport.check;
import static lox.TestSupport.checkEquals;
import static lox.TestSupport.checkThrows;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Locale;

public final class FileTest {
    public static void main(String[] args) throws IOException {
        File tmp = File.createTempFile("lox-rt-file-test", ".txt");
        tmp.deleteOnExit();
        String path = tmp.getAbsolutePath();

        LoxFile writer = LoxFile.open(path, "w");
        check(!LoxOps.equal(writer.getMethod("write"), writer.getMethod("write")),
                "two reads of the same file's 'write' method are not Lox-equal");
        File otherTmp = File.createTempFile("lox-rt-file-test-other", ".txt");
        otherTmp.deleteOnExit();
        LoxFile otherWriter = LoxFile.open(otherTmp.getAbsolutePath(), "w");
        check(!LoxOps.equal(writer.getMethod("write"), otherWriter.getMethod("write")),
                "two different files' 'write' method values are not Lox-equal");
        otherWriter.close();
        writer.writeline("first");
        writer.writeline("second");
        writer.write("third-no-newline");
        writer.close();
        checkThrows(writer::read, LoxError.class, "a closed file rejects further reads");

        LoxFile reader = LoxFile.open(path, "r");
        checkEquals("first", reader.readline(), "readline() returns the first line");
        checkEquals("second", reader.readline(), "readline() advances to the next line");
        checkEquals("third-no-newline", reader.readline(), "a trailing line with no newline still counts");
        checkEquals(null, reader.readline(), "readline() at EOF is nil");
        checkThrows(() -> reader.write("x"), LoxError.class, "a read-only file rejects writes");
        reader.close();

        LoxFile reader2 = LoxFile.open(path, "r");
        LoxList lines = (LoxList) reader2.readlines();
        checkEquals(3, lines.elements.size(), "readlines() returns every line");
        reader2.close();

        LoxFile appender = LoxFile.open(path, "a");
        appender.writeline("fourth");
        appender.close();
        LoxFile reader3 = LoxFile.open(path, "r");
        String all = (String) reader3.read();
        check(all.endsWith("fourth\n"), "append mode adds after existing content");
        reader3.close();

        File rwTmp = File.createTempFile("lox-rt-file-rplus", ".txt");
        rwTmp.deleteOnExit();
        LoxFile seed = LoxFile.open(rwTmp.getAbsolutePath(), "w");
        seed.write("seed");
        seed.close();
        LoxFile rw = LoxFile.open(rwTmp.getAbsolutePath(), "r+");
        checkEquals("seed", rw.read(), "r+ can read existing content");
        rw.close();

        checkThrows(() -> LoxFile.open(path, "bogus"), LoxError.class, "open() rejects an invalid mode");

        // PR #97: file I/O must round-trip a raw high byte 1:1, never as
        // a 2-byte UTF-8 sequence — `open()`'s ISO-8859-1 boundary already
        // guaranteed this; the fix was stdout/stdin catching up to it.
        File byteTmp = File.createTempFile("lox-rt-file-bytes", ".bin");
        byteTmp.deleteOnExit();
        LoxFile byteWriter = LoxFile.open(byteTmp.getAbsolutePath(), "w");
        byteWriter.write(String.valueOf((char) 0xE9));
        byteWriter.close();
        checkEquals(1L, byteTmp.length(), "one Lox++ char of value 0xE9 writes exactly one byte");
        LoxFile byteReader = LoxFile.open(byteTmp.getAbsolutePath(), "r");
        String byteBack = (String) byteReader.read();
        byteReader.close();
        checkEquals(1, byteBack.length(), "the byte reads back as one char, not a replacement pair");
        checkEquals((int) 0xE9, (int) byteBack.charAt(0), "the byte round-trips to its exact value");

        // Same call, dispatched through LoxOps.invoke's no-allocation path.
        LoxFile forInvoke = LoxFile.open(rwTmp.getAbsolutePath(), "r");
        checkEquals("seed", LoxOps.invoke(forInvoke, "read", new Object[0]), "invoke() dispatches file methods directly");
        forInvoke.close();

        // Directory opens: on Linux, fopen(dir, "r") succeeds but read(2)
        // fails with EISDIR. LoxFile.open must return a file and defer to
        // read()/readline()/readlines().
        if (System.getProperty("os.name", "").toLowerCase(Locale.ROOT).contains("linux")) {
            Path dirPath = Files.createTempDirectory("lox-rt-file-dir-test");
            try {
                String dir = dirPath.toString();
                LoxFile dirReader = LoxFile.open(dir, "r");
                check(dirReader != null, "open(directory, \"r\") succeeds and returns a file");
                checkEquals("", dirReader.read(), "read() on a directory gives empty string");
                checkEquals("", dirReader.read(), "read() on a directory again gives empty string");
                checkEquals(null, dirReader.readline(), "readline() on a directory gives nil");
                LoxList emptyList = (LoxList) dirReader.readlines();
                checkEquals(0, emptyList.elements.size(), "readlines() on a directory gives an empty list");
                checkThrows(() -> dirReader.write("x"), LoxError.class,
                        "write() on a read-only directory file raises an error");
                checkThrows(() -> dirReader.writeline("x"), LoxError.class,
                        "writeline() on a read-only directory file raises an error");
                dirReader.close();
                checkThrows(dirReader::read, LoxError.class,
                        "read() on a closed directory file raises an error");
                checkThrows(dirReader::readline, LoxError.class,
                        "readline() on a closed directory file raises an error");
                dirReader.close();
                checkThrows(() -> LoxFile.open(dir, "w"), LoxError.class,
                        "open(directory, \"w\") raises an error");
                checkThrows(() -> LoxFile.open(dir, "a"), LoxError.class,
                        "open(directory, \"a\") raises an error");
                checkThrows(() -> LoxFile.open(dir, "r+"), LoxError.class,
                        "open(directory, \"r+\") raises an error");
                checkThrows(() -> LoxFile.open(dir, "bogus"), LoxError.class,
                        "open(directory, \"bogus\") raises an error for invalid mode");

                Path symlinkPath = dirPath.resolveSibling(
                        dirPath.getFileName() + "-link-" + System.nanoTime());
                try {
                    Files.createSymbolicLink(symlinkPath, dirPath);
                    LoxFile symlinkReader = LoxFile.open(symlinkPath.toString(), "r");
                    check(symlinkReader != null, "open(link to directory, \"r\") succeeds");
                    checkEquals("", symlinkReader.read(),
                            "read() on a link to a directory gives empty string");
                    symlinkReader.close();
                } finally {
                    Files.deleteIfExists(symlinkPath);
                }

                String missing = dirPath.resolve("lox-rt-file-dir-missing-" + System.nanoTime() + ".txt")
                        .toString();
                checkThrows(() -> LoxFile.open(missing, "r"), LoxError.class,
                        "open(missing path, \"r\") raises an error even with directory check");
            } finally {
                Files.deleteIfExists(dirPath);
            }
        }

        // Issue #401: NUL byte handling in File.read/readline/readlines
        // Create a temp file with NUL bytes
        File nulTmp = File.createTempFile("lox-rt-file-nul", ".bin");
        nulTmp.deleteOnExit();
        String nulPath = nulTmp.getAbsolutePath();

        // Test read() with NUL bytes
        LoxFile nulWriter = LoxFile.open(nulPath, "w");
        nulWriter.write("A\0BCDEFGH");
        nulWriter.close();

        LoxFile nulReader = LoxFile.open(nulPath, "r");
        String readData = (String) nulReader.read();
        nulReader.close();
        checkEquals(9, readData.length(), "read() returns all 9 bytes including NUL");
        checkEquals('A', readData.charAt(0), "first byte is 'A'");
        checkEquals('\0', readData.charAt(1), "second byte is NUL");
        checkEquals('B', readData.charAt(2), "third byte is 'B'");
        checkEquals('H', readData.charAt(8), "last byte is 'H'");

        // Test readline() with NULs in line (no trailing newline)
        nulWriter = LoxFile.open(nulPath, "w");
        nulWriter.write("line1\0with\0nuls");
        nulWriter.close();

        nulReader = LoxFile.open(nulPath, "r");
        String line1 = (String) nulReader.readline();
        nulReader.close();
        checkEquals(15, line1.length(), "readline() preserves NULs in line without newline");
        checkEquals("line1\0with\0nuls", line1, "readline() content matches");

        // Test readline() with NULs and trailing newline
        nulWriter = LoxFile.open(nulPath, "w");
        nulWriter.write("line1\0with\0nuls\nline2\n");
        nulWriter.close();

        nulReader = LoxFile.open(nulPath, "r");
        line1 = (String) nulReader.readline();
        String line2 = (String) nulReader.readline();
        nulReader.close();
        checkEquals(15, line1.length(), "first line length with NULs");
        checkEquals("line1\0with\0nuls", line1, "first line content");
        checkEquals(5, line2.length(), "second line length");
        checkEquals("line2", line2, "second line content");

        // Test readlines() with multiple lines containing NULs
        nulWriter = LoxFile.open(nulPath, "w");
        nulWriter.write("A\0B\nC\0D\nE\n");
        nulWriter.close();

        nulReader = LoxFile.open(nulPath, "r");
        LoxList nulLines = (LoxList) nulReader.readlines();
        nulReader.close();
        checkEquals(3, nulLines.elements.size(), "readlines() returns 3 lines");
        checkEquals("A\0B", nulLines.elements.get(0), "first line: A\\0B");
        checkEquals("C\0D", nulLines.elements.get(1), "second line: C\\0D");
        checkEquals("E", nulLines.elements.get(2), "third line: E");

        System.exit(TestSupport.finish("FileTest"));
    }
}

File: runtime/clr/test/FileTest.cs
using System.IO;
using Lox;

namespace LoxRuntimeTests;

public static class FileTest {
    public static int Run() {
        var t = new TestSupport();

        string path = Path.Combine(Path.GetTempPath(), $"lox-rt-file-test-{System.Guid.NewGuid():N}.txt");
        try {
            LoxFile writer = LoxFile.Open(path, "w");
            t.Check(ReferenceEquals(writer.GetMethod("write"), writer.GetMethod("write")),
                "getMethod caches: repeated access returns the same object");
            writer.Writeline("first");
            writer.Writeline("second");
            writer.Write("third-no-newline");
            writer.Close();
            t.CheckThrows(() => writer.Read(), typeof(LoxError), "a closed file rejects further reads");

            LoxFile reader = LoxFile.Open(path, "r");
            t.CheckEquals("first", reader.Readline(), "readline() returns the first line");
            t.CheckEquals("second", reader.Readline(), "readline() advances to the next line");
            t.CheckEquals("third-no-newline", reader.Readline(), "a trailing line with no newline still counts");
            t.CheckEquals(null, reader.Readline(), "readline() at EOF is nil");
            t.CheckThrows(() => reader.Write("x"), typeof(LoxError), "a read-only file rejects writes");
            reader.Close();

            LoxFile reader2 = LoxFile.Open(path, "r");
            var lines = (LoxList)reader2.Readlines();
            t.CheckEquals(3, lines.Elements.Count, "readlines() returns every line");
            reader2.Close();

            LoxFile appender = LoxFile.Open(path, "a");
            appender.Writeline("fourth");
            appender.Close();
            LoxFile reader3 = LoxFile.Open(path, "r");
            var all = (string)reader3.Read();
            t.Check(all.EndsWith("fourth\n"), "append mode adds after existing content");
            reader3.Close();

            string rwPath = Path.Combine(Path.GetTempPath(), $"lox-rt-file-rplus-{System.Guid.NewGuid():N}.txt");
            try {
                LoxFile seed = LoxFile.Open(rwPath, "w");
                seed.Write("seed");
                seed.Close();
                LoxFile rw = LoxFile.Open(rwPath, "r+");
                t.CheckEquals("seed", rw.Read(), "r+ can read existing content");
                rw.Close();

                // Same call, dispatched through LoxOps.Invoke's no-allocation path.
                LoxFile forInvoke = LoxFile.Open(rwPath, "r");
                t.CheckEquals("seed", LoxOps.Invoke(forInvoke, "read", System.Array.Empty<object>()),
                    "invoke() dispatches file methods directly");
                forInvoke.Close();
            } finally {
                File.Delete(rwPath);
            }

            t.CheckThrows(() => LoxFile.Open(path, "bogus"), typeof(LoxError), "open() rejects an invalid mode");

            // r+ must behave like C fopen(path, "r+"): fail when the path
            // does not exist, and - just as importantly - not create it.
            string missingPath = Path.Combine(Path.GetTempPath(), $"lox-rt-file-rplus-missing-{System.Guid.NewGuid():N}.txt");
            t.CheckThrows(() => LoxFile.Open(missingPath, "r+"), typeof(LoxError),
                "open(missing path, \"r+\") fails instead of creating the file");
            t.Check(!File.Exists(missingPath), "open(missing path, \"r+\") leaves no file behind even after it fails");

            // The native VM keeps one ObjNative per method name in a
            // class-wide table shared by every ObjFile, so
            // `f1.write == f2.write` is true there even though f1 and f2
            // are different files - through LoxOps.Equal's Lox-level
            // notion of equality, not C# reference equality.
            string otherPath = Path.Combine(Path.GetTempPath(), $"lox-rt-file-other-{System.Guid.NewGuid():N}.txt");
            try {
                LoxFile other = LoxFile.Open(otherPath, "w");
                t.Check(LoxOps.Equal(writer.GetMethod("write"), other.GetMethod("write")),
                    "two different files' 'write' method values are Lox-equal");
                t.Check(!LoxOps.Equal(writer.GetMethod("write"), other.GetMethod("close")),
                    "two different method names on files are not Lox-equal");
                other.Close();
            } finally {
                File.Delete(otherPath);
            }

            // File I/O must round-trip a raw high byte 1:1, never as a
            // 2-byte UTF-8 sequence - open()'s Latin1 boundary guarantees
            // this.
            string bytePath = Path.Combine(Path.GetTempPath(), $"lox-rt-file-bytes-{System.Guid.NewGuid():N}.bin");
            try {
                LoxFile byteWriter = LoxFile.Open(bytePath, "w");
                byteWriter.Write(((char)0xE9).ToString());
                byteWriter.Close();
                t.CheckEquals(1L, new FileInfo(bytePath).Length, "one Lox++ char of value 0xE9 writes exactly one byte");
                LoxFile byteReader = LoxFile.Open(bytePath, "r");
                var byteBack = (string)byteReader.Read();
                byteReader.Close();
                t.CheckEquals(1, byteBack.Length, "the byte reads back as one char, not a replacement pair");
                t.CheckEquals((int)0xE9, (int)byteBack[0], "the byte round-trips to its exact value");
            } finally {
                File.Delete(bytePath);
            }
        } finally {
            File.Delete(path);
        }

        // Directory-opening tests: on Linux, fopen(dir, "r") succeeds but
        // read(2) fails with EISDIR. Here, LoxFile.Open must return a LoxFile
        // and defer the error to read()/readline()/readlines() methods.
        // These tests guard against future changes to Directory.Exists.
        if (OperatingSystem.IsLinux()) {
            string dirPath = Path.Combine(Path.GetTempPath(), $"lox-rt-file-dir-test-{System.Guid.NewGuid():N}");
            Directory.CreateDirectory(dirPath);
            try {
                // Test 1: open(dir, "r") succeeds and gives a LoxFile.
                LoxFile dirReader = LoxFile.Open(dirPath, "r");
                t.Check(dirReader != null, "open(directory, \"r\") succeeds and returns a LoxFile");

                // Test 2 & 3: read() returns "" each time.
                t.CheckEquals("", dirReader.Read(), "read() on a directory gives empty string");
                t.CheckEquals("", dirReader.Read(), "read() on a directory again gives empty string");

                // Test 4: readline() gives null.
                t.CheckEquals(null, dirReader.Readline(), "readline() on a directory gives nil");

                // Test 5: readlines() gives an empty list.
                var emptyList = (LoxList)dirReader.Readlines();
                t.CheckEquals(0, emptyList.Elements.Count, "readlines() on a directory gives an empty list");

                // Test 6: write() fails (file is not open for writing).
                t.CheckThrows(() => dirReader.Write("x"), typeof(LoxError),
                    "write() on a read-only directory file raises LoxError");

                // Test 7: after close(), read() fails (file is closed).
                dirReader.Close();
                t.CheckThrows(() => dirReader.Read(), typeof(LoxError),
                    "read() on a closed directory file raises LoxError");

                // Test 8: modes "w", "a", "r+" still raise LoxError on a directory.
                t.CheckThrows(() => LoxFile.Open(dirPath, "w"), typeof(LoxError),
                    "open(directory, \"w\") raises LoxError");
                t.CheckThrows(() => LoxFile.Open(dirPath, "a"), typeof(LoxError),
                    "open(directory, \"a\") raises LoxError");
                t.CheckThrows(() => LoxFile.Open(dirPath, "r+"), typeof(LoxError),
                    "open(directory, \"r+\") raises LoxError");

                // Test 9: invalid mode still fails.
                t.CheckThrows(() => LoxFile.Open(dirPath, "bogus"), typeof(LoxError),
                    "open(directory, \"bogus\") raises LoxError for invalid mode");

                // Test 10: symbolic link to directory behaves the same.
                string symlinkPath = Path.Combine(Path.GetTempPath(), $"lox-rt-file-symlink-dir-{System.Guid.NewGuid():N}");
                try {
                    File.CreateSymbolicLink(symlinkPath, dirPath);
                    LoxFile symlinkReader = LoxFile.Open(symlinkPath, "r");
                    t.Check(symlinkReader != null, "open(symlink to directory, \"r\") succeeds");
                    t.CheckEquals("", symlinkReader.Read(), "read() on a symlink to directory gives empty string");
                    symlinkReader.Close();
                } finally {
                    if (File.Exists(symlinkPath)) {
                        File.Delete(symlinkPath);
                    }
                }

                // Test 11: missing path still fails with mode "r".
                string missingPath = Path.Combine(Path.GetTempPath(), $"lox-rt-file-dir-missing-{System.Guid.NewGuid():N}");
                t.CheckThrows(() => LoxFile.Open(missingPath, "r"), typeof(LoxError),
                    "open(missing path, \"r\") raises LoxError even with directory check in place");
            } finally {
                Directory.Delete(dirPath);
            }
        }

        return t.Finish("FileTest");
    }
}
